File: demo1/src/main/java/com/hery/flink/java/StreamingJob.java
package com.hery.flink.java;

import com.hery.flink.java.utils.WordCountData;
import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.common.functions.MapFunction;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;

/**
 * @Date 2020/6/23 17:49
 * @Created by hery
 * @Description flink 流处理
 */
public class StreamingJob {

    public static void main(String[] args) throws Exception {
        // set up the streaming execution environment
        final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        DataStream<String> text = null;
        text = env.fromElements(WordCountData.WORDS);

        DataStream<Tuple2<String, Integer>> counts =
                // split up the lines in pairs (2-tuples) containing: (word,1)
                text.flatMap(new Tokenizer()).map(new MapFunction<Tuple2<String, Integer>, Tuple2<String, Integer>>() {
                    @Override
                    public Tuple2<String, Integer> map(Tuple2<String, Integer> tuple2) throws Exception {

                        return tuple2;
                    }
                })
                        // group by the tuple field "0" and sum up tuple field "1"
                        .keyBy(0).sum(1);
        counts.print();
        // execute program
        env.execute("Flink StreamingJob");
    }

    public static final class Tokenizer implements FlatMapFunction<String, Tuple2<String, Integer>> {

        @Override
        public void flatMap(String value, Collector<Tuple2<String, Integer>> out) {
            // normalize and split the line
            String[] tokens = value.toLowerCase().split("\\W+");

            // emit the pairs
            for (String token : tokens) {
                if (token.length() > 0) {
                    out.collect(new Tuple2<>(token, 1));
                }
            }
        }
    }
}

File: demo1/src/main/scala/com/hery/flink/scala/StreamingJob.scala
package com.hery.flink.scala

import com.hery.flink.java.utils.WordCountData
import org.apache.flink.streaming.api.scala._


/**
 * @Date 2020/7/1 16:50
 * @Created by hery
 * @Description flink scala 流处理
 */
object StreamingJob {
  def main(args: Array[String]): Unit = {

    // 设置环境
    val env = StreamExecutionEnvironment.getExecutionEnvironment

    // 数据读取
    val text = env.fromElements(WordCountData.WORDS: _*)

    // 数据处理
    val counts: DataStream[(String, Int)] = text
      // split up the lines in pairs (2-tuples) containing: (word,1)
      .flatMap(_.toLowerCase.split("\\W+"))
      .filter(_.nonEmpty)
      .map((_, 1))
      // group by the tuple field "0" and sum up tuple field "1"
      .keyBy(0)
      .sum(1)

    // 数据输出
    counts.print()
    // 执行程序
    env.execute("StreamingJob WordCount")
  }
}
